pg_bm25/src/tokenizers/lindera.rs

```diff
@@ -0,0 +1,325 @@
+/*
+ *
+ * IMPORTANT NOTICE:
+ * This file has been copied from Quickwit, an open source project, and is subject to the terms
+ * and conditions of the GNU Affero General Public License (AGPL) version 3.0.
+ * Please review the full licensing details at <http://www.gnu.org/licenses/>.
+ * By using this file, you agree to comply with the AGPL v3.0 terms.
+ *
+ */
+
+use lindera_core::mode::Mode;
+use lindera_dictionary::{load_dictionary_from_config, DictionaryConfig, DictionaryKind};
+use lindera_tokenizer::token::Token as LinderaToken;
+use lindera_tokenizer::tokenizer::Tokenizer as LinderaTokenizer;
+use once_cell::sync::Lazy;
+use tantivy::tokenizer::{SimpleTokenStream, SimpleTokenizer, Token, TokenStream, Tokenizer};
+use whichlang::{detect_language, Lang};
+
+/// Mandarin chinese tokenizer.
+static CMN_TOKENIZER: Lazy<LinderaTokenizer> = Lazy::new(|| {
+    let cmn_dictionary_config = DictionaryConfig {
+        kind: Some(DictionaryKind::CcCedict),
+        path: None,
+    };
+    let cmn_dictionary = load_dictionary_from_config(cmn_dictionary_config)
+        .expect("Lindera `CcCedict` dictionary must be present");
+    LinderaTokenizer::new(cmn_dictionary, None, Mode::Normal)
+});
+
+/// Japanese tokenizer.
+static JPN_TOKENIZER: Lazy<LinderaTokenizer> = Lazy::new(|| {
+    let jpn_dictionary_config = DictionaryConfig {
+        kind: Some(DictionaryKind::IPADIC),
+        path: None,
+    };
+    let jpn_dictionary = load_dictionary_from_config(jpn_dictionary_config)
+        .expect("Lindera `IPAD` dictionary must be present");
+    LinderaTokenizer::new(jpn_dictionary, None, Mode::Normal)
+});
+
+/// Korean tokenizer.
+static KOR_TOKENIZER: Lazy<LinderaTokenizer> = Lazy::new(|| {
+    let kor_dictionary_config = DictionaryConfig {
+        kind: Some(DictionaryKind::KoDic),
+        path: None,
+    };
+    let kor_dictionary = load_dictionary_from_config(kor_dictionary_config)
+        .expect("Lindera `KoDic` dictionary must be present");
+    LinderaTokenizer::new(kor_dictionary, None, Mode::Normal)
+});
+
+/// Multilanguage tokenizer that uses the `whichlang` to detect the language of the text
+/// and uses the appropriate tokenizer for the detected language:
+/// - lindera for Chinese, Japanese, and Korean.
+/// - Quickwit's default tokenizer for other languages.
+/// It is possible to bypass the language detection by prefixing the text with the language code
+/// followed by a colon. For example, `KOR:일본입니다` will be tokenized by the english tokenizer.
+/// Current supported prefix are:
+/// - `KOR:` for Korean tokenizer
+/// - `JPN:` for Japanese tokenizer
+/// - `CMN:` for Chinese tokenizer
+/// - `ENG:` for Quickwit's default tokenizer
+#[derive(Clone, Default)]
+pub struct MultiLangTokenizer {
+    default_tokenizer: SimpleTokenizer,
+    token: Token,
+}
```

Comment:
We'll probably want a docs entry explaining how to do this. The comments are great, but we should probably have some clear examples of queries that use this syntax.

Is this what we're expecting?
```sql
'description:JPN:すもももももももものうち:::fuzzy_fields=description&fast=true'
```

Reply:
Is it only these three additional languages that we're supporting with this? If that's the case, I don't think we should should necessarily offload the language parsing to underlying libraries. Let's just parametrize ` ParadeTokenizer::Lindera` to take a language, like we've done with ngrams.

I think another good reason to do this is that we don't want to specify tokenizer parameters at query time. If the index is tokenized with JPN, it's ambiguous what will happen if you query with KOR. We shouldn't even allow the user to make this mistake, as tokenizer config should happen at index build time.

Reply:
I kind of agree. I copy-pasted this file from Quickwit. Happy to take your suggestion. 

Also WRT to the JPN:* syntax, I believe this only makes sense for the text being indexed. My understanding is that Tantivy will tokenize the query the same way that the queried field is tokenized.

Reply:
Done!